Theory_assignment_two/assignment_2.md

```diff
@@ -0,0 +1,453 @@
+- Inputs and outputs for every module
+- Listing of error messages
+
+
+4) Implementation
+
+This is where the developers code using chosen programming languages. Tasks are devided into units or modules and assigned to different developers. The implementation phase os the longest phase of the SDLC process.
+
+
+5) Testing
+
+Once the software is completed it is deployed in the testing environment. The testing team check the functionality of the entire system. This is to verify that the entire application works to the customers needs. Testing involves:
+
+-Unit Testing
+-Module Testing
+-Sub-system Testing
+-System Testing
+-Acceptance Testing
+
+
+6) Evaluation
+
+Once a system has been delivered to the end user/client it undergoes a post-implementation review. The delivered system should then fully evaluate on the basis of its
+-effectiveness, usability and maintainability
+
+
+7) Maintenance 
+
+Once the software is deployed and is used by customers.
+
+The software will be maintained by fixing bugs, undergoing upgrades and enhancement through the addition of new features.
+
+
+## 6.What's the difference between agile and waterfall? 
+
+|     Agile                                                      |     Waterfall                                    |   |   |   |
+|----------------------------------------------------------------|--------------------------------------------------|---|---|---|
+|     Multiple lifecycle phases can run in   parallel            |     Isolated phases that flow into each other    |   |   |   |
+|     Continuous iteration of development and   testing          |     Structured software development process      |   |   |   |
+|     Not as useful for small development   projects             |     Better suited for smaller sized projects     |   |   |   |
+|     Incremental approach                                       |     Sequential design process                    |   |   |   |
+|     Testing occurs concurrently with software   development    |     Testing comes after the building phase       |   |   |   |
+|                                                                |                                                  |   |   |   |
+
+
+
+
+
+
+
+## 7. What is a reduced function used for? 
+
+Reduce() is used for applying a function to an iterable and reducing it to a single cumulative value. It performs function on first two elements and repeats process until 1 value remains.
+
+```python
+#example of a reduced function
+import functools
+
+letters = ["H","E","L","L","O"]
+word = functools.reduce(lambda x, y, :x + y,letters)
+print(word)
+```
+
+## 8. How does merge sort work 
+
+A merge sort is a sorting algorithm that is based on divide and conquer. An array is divided into two. The subarray are then divided again and again until the arrays only have one element each.  Then pairs of one-element arrays are sorted and are merged with other pairs until the initial array is sorted.
```

Comment:
5/10 The idea of the question is not necessary to write the full code.
We want to see how a student can understand the algorithm. The expected answer should be a description of how merge sort work and a step-by-step example of each iteration. A visual sort also can be helpful to see each iteration during sorting.